### PR TITLE
Add configuration

### PR DESCRIPTION
**Pull request by kychum**
Adds basic configuration and pulls settings out to prepare for hot reloading

### CODE REVIEW

File: src/AudioSettings.h [hunk 1]
```diff
@@ -0,0 +1,18 @@
+#pragma once
+
+namespace AudioSettings {
+    // Sound sampling frequency in Hz. Change to "22050" if too slow.
+    int MIX_FREQ_HZ = 44100;
+    // Number of sound channels for output. 1 = Mono, 2 = Stereo. Indepent from mixing channels.
+    int MIX_NUM_CHANNELS = 2;
+    // Bytes used per output sample
+    int MIX_CHUNK_SIZE = 4096;
```
Comment (enochtsang) [inline src/AudioSettings.h]: The variables should be camelCase if they're not constants.

File: src/AudioSettings.h [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+    // Information for loading sound assets. First value is the key, second is the path.
+    std::vector<std::pair<const SoundAsset, const char*>> SOUND_ASSETS_INFO = {
+        {SoundAsset::BEAT, "assets/audio/beat.wav"}
+    };
+    // Information for loading music assets. First value is the key, second is the path.
+    std::vector<std::pair<const MusicAsset, const char*>> MUSIC_ASSETS_INFO = {
```
Comment (enochtsang) [inline src/AudioSettings.h]: Instead of using `const char*`, string would be better since it has more functionality. I know it wasn't you since you copied it from somewhere, but can you change it anyways? :)

File: src/Event.h [hunk 3]
```diff
@@ -4,7 +4,7 @@
 #include <unordered_map>
 #include <string>
 
-enum class EventType {LOAD_EVENT, DUMMY_EVENT, KEYPRESS_EVENT, EVENT_COUNT};
+enum class EventType {LOAD_EVENT, DUMMY_EVENT, KEYPRESS_EVENT, SETTINGS_EVENT, EVENT_COUNT};
```
Comment (enochtsang) [inline src/Event.h]: `SETTINGS_EVENT` isn't a good name because `SETTINGS` isn't an action. `RELOAD_SETTINGS_EVENT` would be better.

File: src/AudioSettings.h [hunk 4]
```diff
@@ -0,0 +1,18 @@
+#pragma once
+
+namespace AudioSettings {
```
Comment (enochtsang) [inline src/AudioSettings.h]: Just putting this in a namespace make these values global. It'd be better to make this a struct that get explicitly instantiated somewhere.

File: src/InputSettings.h [hunk 5]
```diff
@@ -0,0 +1,12 @@
+#pragma once
+#include "SettingType.h"
+#include<map>
+
+namespace InputSettings {
+    const int MAX_PLAYERS = 4; // Maximum number of players in a game
+    extern int DEADZONE; // Minimum range of displacement for joystick before reading event
+
+    const std::map<std::string, std::pair<const SettingType, void* const>> CONFIG_MAPPING = {
```
Comment (enochtsang) [inline src/InputSettings.h]: A map of settings isn't necessary imo, adds complexity where it really isn't necessary. Also it seems like some settings are a special case that are not in the map while others are not, doesn't really add up.

File: src/InputSettings.h [hunk 6]
```diff
@@ -0,0 +1,12 @@
+#pragma once
+#include "SettingType.h"
+#include<map>
+
+namespace InputSettings {
```
Comment (enochtsang) [inline src/InputSettings.h]: This would do better as a struct as well.

File: src/main.cpp [hunk 7]
```diff
@@ -12,14 +12,17 @@
 #include "InputManager.h"
 #include "RenderingSystem.h"
 
+#include "SettingsSystem.h"
 
+const char* SETTINGS_FILE = "config/config.yaml";
```
Comment (enochtsang) [inline src/main.cpp]: Better to use `string`, less error prone.

File: src/RenderingSystem.cpp [hunk 8]
```diff
@@ -86,32 +86,32 @@ bool RenderingSystem::init_window() {
...
 
     if (window_ == NULL) {
         std::cout << "Window could not be created! SDL Error: " << SDL_GetError() << std::endl;
         return false;
     }
 
 
-    SDL_GL_SetAttribute(SDL_GL_CONTEXT_PROFILE_MASK, SDL_GL_CONTEXT_PROFILE_CORE);
-    SDL_GL_SetAttribute(SDL_GL_CONTEXT_MAJOR_VERSION, 3);
-    SDL_GL_SetAttribute(SDL_GL_CONTEXT_MINOR_VERSION, 2);
```
Comment (enochtsang) [inline src/RenderingSystem.cpp]: Why was this file changed?

File: src/main.cpp [hunk 9]
```diff
@@ -53,6 +56,7 @@ int main(int argc, char* args[]) {
         // Events
         rendering_system.send_events(events);
         input_manager.send_events(events);
+        settings_system.handle_events(events);
```
Comment (enochtsang) [inline src/main.cpp]: The settings_system needs to send_events as well.

File: src/RenderingSystem.cpp (hunk 8 above)
Comment (kychum) [reply src/RenderingSystem.cpp]: On my virtual machine GLEW wasn't getting the right OpenGL versions, failing with "Missing GL version" and eventually crashing with a segfault. From what I gathered, certain SDL_GL_* calls need to be called before or after the context is created to prevent this from happening.

File: src/InputSettings.h (hunk 5 above)
Comment (kychum) [reply src/InputSettings.h]: The way I was thinking about this is that any time we add anything to the namespace/struct that needs to be hot-reloaded, we'd need to update the SettingsSystem as well. One advantage of using the map is that we only need to update the namespace and add a corresponding entry to the map without caring about how the SettingsSystem updates individual values. Unfortunately disadvantage is that we need the extra SettingType enum to determine what kind of value is to be stored in the pointer.

If you think that's still too much complexity I'll go ahead and change things around!

File: src/AudioSettings.h (hunk 4 above)
Comment (kychum) [reply src/AudioSettings.h]: Going to ping @ammunoz here about his thoughts on this too, since he came up with the namespace idea during pair programming. If we plan on storing anything more complicated than a map then I agree that a struct would probably be better.

File: src/InputSettings.h (hunk 5 above)
Comment (enochtsang) [reply src/InputSettings.h]: Yes we would need to add something if we add a configuration value. But if we add a configuration, we probably don't just need to change that anyways, somewhere something is going to use that value and that will need to be updated anyways.

File: src/SettingsSystem.h [hunk 10]
```diff
@@ -0,0 +1,46 @@
+#pragma once
+
+#include<memory>
+#include<yaml-cpp/yaml.h>
```
Comment (enochtsang) [inline src/SettingsSystem.h]: Spaces after the includes

File: src/SettingsSystem.cpp [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+
+    return str + "}";
+}
+
+void SettingsSystem::handle_keypress_event(const Event& e) {
+    int key = e.get_value<int>("key", -1);
+
+    switch (key) {
+        case SDLK_F5:
+            if (reload_settings()) {
```
Comment (enochtsang) [inline src/SettingsSystem.cpp]: What if it fails?

File: src/SettingsSystem.cpp [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+                EventSystem::queue_event( Event( EventType::RELOAD_SETTINGS_EVENT ) );
```
Comment (enochtsang) [inline src/SettingsSystem.cpp]: The other systems should be handling this event in this PR right now.

File: src/SettingsSystem.cpp [hunk 13]
```diff
@@ -0,0 +1,121 @@
+#include "SettingsSystem.h"
+#include "InputSettings.h"
+#include "AudioSettings.h"
+#include <SDL.h>
+
+namespace {
+    using Key = const std::string; // Convenience typedef
```
[the remark this answers is not in the document]
Comment (Pigglesworth) [reply src/SettingsSystem.cpp]: I'd argue that typedefs increase clarity because it's a little more clear how the value will be passed around and used. Usually I'd use the using keywords for things that are more obnoxious but I see no problem with this specific usage.

[the remark this answers is not in the document]
Comment (ammunoz) [reply src/SettingsSystem.cpp]: Just use the old syntax for typedef then. They are equivalent yet avoids "using". Best of both worlds

[the remark this answers is not in the document]
Comment (Pigglesworth) [reply src/SettingsSystem.cpp]: Excellent idea. I 100% approve of such a change.

File: src/SettingsSystem.cpp (hunk 11 above)
Comment (kychum) [reply src/SettingsSystem.cpp]: Ah right, at the moment we just continue with possibly bad settings. I think we should revert back to the original settings if it fails. Does that sound fine?

Comment (enochtsang) [reply src/SettingsSystem.cpp]: yup, a message should get sent too

**Comment by kychum** [comment]
@enochtsang updated. Let me know how the libs.zip should be updated!

File: src/SettingsSystem.h [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+class SettingsSystem : public EventSystem<SettingsSystem> {
+public:
+    SettingsSystem(std::string config_file);
+    bool reload_settings();
+
+    std::shared_ptr<InputSettings> get_input_settings();
+    std::shared_ptr<AudioSettings> get_audio_settings();
+
+    template<typename T>
+    bool load_key(std::vector<std::string> keys, T& value) const {
```
Comment (enochtsang) [inline src/SettingsSystem.h]: This should be in a `_impl` file.

File: src/SettingsSystem.h [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+struct InputSettings;
+struct AudioSettings;
+
+class SettingsSystem : public EventSystem<SettingsSystem> {
+public:
+    SettingsSystem(std::string config_file);
+    bool reload_settings();
+
+    std::shared_ptr<InputSettings> get_input_settings();
+    std::shared_ptr<AudioSettings> get_audio_settings();
```
Comment (enochtsang) [inline src/SettingsSystem.h]: Just return a const reference instead of a shared_ptr.

And they should be const functions.

File: src/SettingsSystem.cpp [hunk 16]
```diff
@@ -0,0 +1,151 @@
...
+    Key KEY_INPUT = "input";
+    Key KEY_INPUT_DEADZONE = "deadzone";
+    Key KEY_INPUT_MAXPLAYERS = "max_players";
+
+    Key KEY_AUDIO = "audio";
+    Key KEY_AUDIO_FREQ = "freq";
+    Key KEY_AUDIO_CHANNELS = "channels";
+    Key KEY_AUDIO_CHUNKSIZE = "chunk_size";
+
+    typedef const std::vector<std::string> Keys;
```
Comment (enochtsang) [inline src/SettingsSystem.cpp]: the std::string should be the `Key` that you have a typedef for above.

File: src/SettingsSystem.cpp [hunk 17]
```diff
@@ -0,0 +1,151 @@
...
+
+    try {
+        head_node_ = YAML::LoadFile(config_file_);
+    } catch (YAML::BadFile e) {
+        std::cerr << "Failed to load settings file " << config_file_ << std::endl;
+        return false;
+    }
+
+    bool success = true;
+    std::shared_ptr<InputSettings> backup_input_settings(new InputSettings(*input_settings_));
```
Comment (enochtsang) [inline src/SettingsSystem.cpp]: This doesn't need to be ptr.

File: src/SettingsSystem.cpp [hunk 18]
```diff
@@ -0,0 +1,151 @@
...
+
+    if (!load_key<int>(KEYS_MAX_PLAYERS, input_settings_->max_players)) {
+        return false;
+    }
+
+    return true;
+}
+
+bool SettingsSystem::reload_audio_settings() {
+    if (!load_key<int>(KEYS_FREQ, audio_settings_->mix_freq_hz)) {
```
Comment (enochtsang) [inline src/SettingsSystem.cpp]: Try this 

```c++
int error = 0;
error += !load_key<int>(KEYS_FREQ, audio_settings_->mix_freq_hz);
...
error += !load_key<int>(KEYS_FREQ, audio_settings_->mix_chunk_size);

return error;
```
